state/state.go

```diff
@@ -34,6 +42,140 @@ type DownloadedObject struct {
 	Object, Filename string
 }
 
+type DynamoDBStater struct {
+	*sync.Mutex
+	Session *session.Session
+	Service string
+}
+
+func NewDynamoDBStater(session *session.Session, service string) *DynamoDBStater {
+	return &DynamoDBStater{
+		Mutex:   &sync.Mutex{},
+		Session: session,
+		Service: service,
+	}
+}
+
+// Used for unmarshaling and adding objects to DynamoDB
+type Record struct {
+	S3Object string
+	Time     time.Time
+}
+
+// list of processed objects
+func (d *DynamoDBStater) processedObjects() (map[string]time.Time, error) {
+	objs := make(map[string]time.Time)
+
+	var records []Record
+	var oldRecords []Record
+
+	svc := dynamodb.New(d.Session)
+	err := svc.ScanPages(&dynamodb.ScanInput{
+		TableName: aws.String("HoneyAWSAccessLogBuckets"),
+	}, func(logs *dynamodb.ScanOutput, last bool) bool {
+		recs := []Record{}
+
+		err := dynamodbattribute.UnmarshalListOfMaps(logs.Items, &recs)
+
+		// break out of function
+		if err != nil {
+			logrus.WithField("error", err).Debug("Failed to unmarshal DynamoDB Scan Items")
+			return false
+		}
+
+		records = append(records, recs...)
+
+		// let's see if any of the time stamps are from the previous day
+		// if so, breakout by returning false
+		// it's okay if we include a couple objects from the previous
+		// day
+		for _, rec := range records {
+			// delete the item from dynamodb so we don't grow
+			// infinitely
+			if time.Since(rec.Time) > BackfillInterval {
+				logrus.WithField("item", rec.S3Object).Debug("Deleting object from DynamoDB table")
+				oldRecords = append(oldRecords, rec)
+			}
+		}
+		return true
+	})
+
+	if err != nil {
+		return objs, fmt.Errorf("Error scanning DynamoDB, %v", err)
+	}
+
+	for _, record := range records {
+		objs[record.S3Object] = record.Time
+	}
+
+	// cleanup old records (older than Backfill)
+	for _, old := range oldRecords {
+		// need to remarshal object
+		oldObj, err := dynamodbattribute.MarshalMap(old)
+		input := &dynamodb.DeleteItemInput{
+			Key:       oldObj,
+			TableName: aws.String(DynamoTableName),
+		}
+		_, err = svc.DeleteItem(input)
```

Comment:
Why do this here instead of in `SetProcessed` ? I wouldn't expect a call just to read the objects (`ProcessedObjects`) to delete data

Reply:
When we do `DeleteItem`, the specific item you're deleting is required - we already have a list of all those items available to use from records when we do an initial scan, so it made sense to just filter that list slightly and reuse stuff already pulled down

The alternative would be to do a `Query` within `SetProcessed` that has some predicate function that filters based on dates, which would result in another roundtrip to DynamoDB every time `SetProcessed` is called (which happens more frequently than `ProcessedObjects` - at least I think so)

It would be nice to do within SetProcessed, because we do that for the file stater, but I didn't want to have to do another roundtrip query!

Reply:
Ah interesting, I just found https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/howitworks-ttl.html so we might not have to delete at all if we set that to 1 hour. What do you think about that idea?

Reply:
Whoa, nice find! This seems like a great avenue to go down - although a thing to consider is the enabling TTL is done during table creation. I think it'd be a good idea (and discussing this with @maplebed) to set TTL to be 7d and using `BackfillInterval` to monitor when we should stop scanning S3 logs. It'd also be nice if we could have `BackfillInterval` be a flag that users can set, with the default still being an hour. This way they can set it up to a max of 7d, or 30 minutes, or whatever.

Reply:
Nice! 7 days feels a bit long to me, but I'm curious what Ben thinks.

Reply:
Ah, I can see where you're going with respect to configurable BackfillInterval though.

Reply:
We discussed this and 7 days seems quite alright, the storage space shouldn't be too bad (likely fitting in free tier)!